route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -41,190 +41,238 @@ namespace route_following_plugin
+    }
+
+    std::vector<cav_msgs::Maneuver> RouteFollowingPlugin::route_cb(lanelet::routing::LaneletPath route_shortest_path){
+        std::vector <cav_msgs::Maneuver> maneuvers;
+        //This function calculates the maneuver plan every time the route is set
+        ROS_DEBUG_STREAM("New route created");
+        //Go through entire route - identify lane changes and fill in the spaces with lane following
+        auto nearest_lanelets = lanelet::geometry::findNearest(wm_->getMap()->laneletLayer, current_loc_, 10);   //Return 10 nearest lanelets
+        if(nearest_lanelets.size() == 0)
         {
             ROS_WARN_STREAM("Cannot find any lanelet in map!");
-            return true;
+            return maneuvers; 
         }
-        
-        auto shortest_path = wm_->getRoute()->shortestPath();
 
-        lanelet::ConstLanelet current_lanelet;
-        int last_lanelet_index = -1;
-        for (auto llt : current_lanelets)
+        //Check if current position is on shortest path, if not on shortest path return empty maneuver  
+        int current_lanelet_index = -1;
+        for (auto llt : nearest_lanelets)
         {
-            if (boost::geometry::within(current_loc, llt.second.polygon2d()))
+            if (boost::geometry::within(current_loc_, llt.second.polygon2d()))
             {
-                int potential_index = findLaneletIndexFromPath(llt.second.id(), shortest_path);
+                int potential_index = findLaneletIndexFromPath(llt.second.id(), route_shortest_path);
                 if (potential_index != -1)
                 {
-                    last_lanelet_index = potential_index;
-                    current_lanelet = shortest_path[last_lanelet_index];
+                    current_lanelet_index = potential_index;
                     break;
                 }
             }
         }
-        if(last_lanelet_index == -1)
+        if(current_lanelet_index == -1)
         {
-            ROS_ERROR_STREAM("Current position is not on the shortest path! Returning an empty maneuver");
-            return true;
+            ROS_WARN_STREAM("Current position is not on the shortest path! Returning an empty maneuver");
         }
-        double current_progress = wm_->routeTrackPos(current_loc).downtrack;
-        double speed_progress = current_speed_;
-        ros::Time time_progress = ros::Time::now();
-        double target_speed=findSpeedLimit(current_lanelet);   //get Speed Limit
 
-        double total_maneuver_length = current_progress + mvr_duration_ * target_speed;
-        double route_length=  wm_->getRouteEndTrackPos().downtrack; 
-        total_maneuver_length = std::min(total_maneuver_length, route_length);
+        double route_length = wm_->getRouteEndTrackPos().downtrack;
+        double prev_mvr_end_speed = 0.0;
+        //Find lane changes in path - up to the second to last lanelet in path (till lane change is possible)
+        for(int i = 0; i < route_shortest_path.size() - 1; ++i){
```

Comment:
Use size_t when defining iterators for for loops. Sonar cloud will throw a warning over this.